catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionContainerImpl.java

```diff
@@ -0,0 +1,396 @@
+        CachedSubscription cachedSubscription = doDelete(subscriptionId);
+        return cachedSubscription.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException("Could not get subscription. "));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized boolean contains(SubscriptionIdentifier identifier) {
+        return get(identifier) != null;
+    }
+
+    /**
+     * Called when the OSGi framework registers a factory, typically from the container's client. It
+     * could be an endpoint, or any other subscription provider.
+     *
+     * @param factory the {@link SubscriptionFactory} that was registered and is now available.
+     */
+    public synchronized void bindFactory(SubscriptionFactory factory) {
+        if (factory == null) {
+            LOGGER.debug("Subscription container binding was given a null factory. ");
+            return;
+        }
+
+        if (factories.containsKey(factory.getTypeName())) {
+            throw new SubscriptionStoreException(
+                    "Duplicate factory registered. This is an API client error");
+        }
+
+        factories.put(factory.getTypeName(), factory);
+        LOGGER.debug("SubscriptionFactory registered for {}", factory.getTypeName());
+
+        StreamSupport.stream(subscriptionsCache.spliterator(), false)
+                .map(Cache.Entry::getValue)
+                .filter(CachedSubscription::isNotRegistered)
+                .filter(sub -> sub.isType(factory.getTypeName()))
+                .forEach(sub -> sub.registerSubscription(factory));
+    }
+
+    /**
+     * Called when the OSGi framework unregisters a factory.
+     *
+     * @param factory the {@link SubscriptionFactory} that was unregistered.
+     * @throws SubscriptionStoreException if an unexpected registration tried to occur.
+     */
+    public synchronized void unbindFactory(SubscriptionFactory factory) {
+        if (factory == null) {
+            LOGGER.debug("Subscription container unbinding was given a null factory. ");
+            return;
+        }
+
+        SubscriptionFactory removedFactory = factories.remove(factory.getTypeName());
+        if (removedFactory == null) {
+            throw new SubscriptionStoreException(format(
+                    "Binding synchronization is wrong. No factory [%s] exists to unbind. ",
+                    factory.getTypeName()));
+        }
+        LOGGER.debug("SubscriptionFactory removed for {}", factory.getTypeName());
+    }
+
+    /**
+     * Helper method for core insert logic.
+     */
+    private void doInsert(Subscription subscription, SubscriptionMetadata metadata) {
+        CachedSubscription cachedSubscription = createCachedSubscription(metadata);
+        cachedSubscription.registerSubscription(subscription);
+        try {
+            subscriptionsCache.put(metadata.getId(), cachedSubscription);
+        } catch (CacheException e) {
```

Comment:
Yes. NullPointer, IllegalState, and ClassCast could occur. See:
https://static.javadoc.io/javax.cache/cache-api/1.0.0/javax/cache/Cache.html#put(K, V)
I would argue anything short of a cache exception is programmer error and _should_ be allowed to propagate.